eq-author/src/components/QCodeContext/index.js

```diff
@@ -1,7 +1,92 @@
-import { createContext, useContext } from "react";
+import React, { createContext, useContext } from "react";
+import PropTypes from "prop-types";
+import CustomPropTypes from "custom-prop-types";
+
+import { RADIO } from "constants/answer-types";
 
 export const QCodeContext = createContext();
 
+export const getAllPages = ({ sections }) =>
+  sections?.flatMap(({ folders }) => folders?.flatMap(({ pages }) => pages));
+
+export const getFlattenedAnswerRows = (questionnaire) =>
+  getAllPages(questionnaire)
+    ?.filter(({ answers }) => answers?.length)
+    ?.flatMap(({ title, alias, answers }) =>
+      answers
+        .flatMap((answer) =>
+          [
+            answer,
+            ...(answer.options?.map((option) => ({
+              ...option,
+              type: answer.type === RADIO ? "RadioOption" : "CheckboxOption",
+              option: true,
+            })) ?? []),
+            answer.mutuallyExclusiveOption && {
+              ...answer.mutuallyExclusiveOption,
+              type: "MutuallyExclusiveOption",
+              option: true,
+            },
+            answer.secondaryLabel && {
+              id: answer.id,
+              label: answer.secondaryLabel,
+              qCode: answer.secondaryQCode,
+              type: answer.type,
+              validationErrorInfo: answer.validationErrorInfo,
+              secondary: true,
+            },
+          ].filter(Boolean)
+        )
+        .map((answer, index) => ({
+          ...answer,
+          questionTitle: title,
+          questionShortCode: alias,
+          nested: index > 0,
+        }))
+    );
```

Comment:
This is quite a lot of chained together code. It would be easier to follow if you made it a traditional function and then broke it up into smaller chunks inside.

You could also likely combine the `.filter()` and `.flatMap()` (lines 14 and 15): you could check if there are any `answers` as you're mapping, enabling you to remove the `.filter()`.

Reply:
Haha, yes - I went into a chaining frenzy when I wrote that :D

I'll have a wee play with it and see if I can make it more comprehensible

Reply:
I've broken this down a bit and put a few more comments in... although admittedly the main structure for processing answers remains the same. I did a more imperative-style rewrite but it looked even uglier than the original. See what you think.

Good spot w/ the `useMemo` dep too, thanks - comitted that one.

Reply:
You're welcome!! I'm of two minds with the revised edition. It does look more ugly, but I can understand it a little more. I think that might be because of your comments, 'though. Maybe we could swap back to the chaining edition, but add your comments in from the revised one?